File: go/x3_oop_alien.go
package main

import . "g2d"

var arena = NewArena(Point{480, 360})
var a1 = NewAlien(arena, Point{40, 40})
var a2 = NewAlien(arena, Point{80, 80})

type Alien struct {
    arena      *Arena
    x, y, w, h int
    xmin, xmax int
    dx, dy     int
}

func NewAlien(arena *Arena, pos Point) *Alien {
    a := &Alien{arena, pos.X, pos.Y, 20, 20, pos.X, pos.X+150, 5, 5}
    arena.Add(a)
    return a
}

func (a *Alien) Move() {
    if a.xmin <= a.x+a.dx && a.x+a.dx <= a.xmax {
        a.x += a.dx
    } else {
        a.dx = -a.dx
        a.y += a.dy
    }
}

func (a *Alien) Position() Point {
    return Point{a.x, a.y}
}

func (a *Alien) Size() Point {
    return Point{a.w, a.h}
}

func (a *Alien) Symbol() Point {
    return Point{0, 0}
}

func (a *Alien) Collide(other Actor) {
}

func tick() {
    ClearCanvas()
    arena.MoveAll()
    for _, actor := range arena.Actors()  {
        FillRect(actor.Position(), actor.Size())
    }
}

func main() {
    InitCanvas(arena.Size())
    MainLoop(tick)
}

File: examples/x3_oop_alien.py
#!/usr/bin/env python3
'''
@author  Michele Tomaiuolo - http://www.ce.unipr.it/people/tomamic
@license This software is free - http://www.gnu.org/licenses/gpl.html
'''

import g2d, random
from actor import Actor, Arena

class Alien(Actor):
    def __init__(self, pos):
        self._x, self._y = pos
        self._w, self._h = 20, 20
        self._xmin, self._xmax = self._x, self._x + 150
        self._dx, self._dy = 5, 5

    def act(self, arena):
        if self._xmin <= self._x + self._dx <= self._xmax:
            self._x += self._dx
        else:
            self._dx = -self._dx
            self._y += self._dy

    def pos(self):
        return self._x, self._y

    def size(self):
        return self._w, self._h

    def sprite(self):
        return 0, 0

    def collide(self, other, arena):
        if isinstance(other, Bullet):
            arena.kill(self)


class Bullet(Actor):
    def __init__(self, x0: int):
        self._w, self._h = 5, 10
        self._x, self._y = x0, arena.size()[1] - self._h
        self._dy = -5

    def act(self, arena):
        self._y += self._dy
        if self._y < 0:
            arena.kill(self)

    def collide(self, other, arena):
        if isinstance(other, Alien):
            arena.kill(self)

    def pos(self):
        return self._x, self._y

    def size(self):
        return self._w, self._h

    def sprite(self):
        return 0, 0


def tick():
    g2d.clear_canvas()
    if random.randrange(50) == 0:
        arena.spawn(Bullet(random.randrange(arena.size()[0])))
    arena.tick()
    for a in arena.actors():
        g2d.fill_rect(a.pos(), a.size())

def main():
    global arena
    arena = Arena((320, 240))
    arena.spawn(Alien((40, 40)))
    arena.spawn(Alien((80, 80)))
    g2d.init_canvas(arena.size())
    g2d.main_loop(tick)

main()
